### ISSUE TITLE
Need interim truth-matching solution/downstream EM shower rollup

### ISSUE THREAD

**Issue by wesketchum** [SBNSoftware/sbndcode#230]
Because we are still waiting on the LArSoft patch for #202 , we need an interim solution for this for SBND (ICARUS did not make the jump to the new LArG4 for this release, so it shouldn't affect them).

As discussed in the Release Management meeting, we'll move ahead without this for this week/production testing, but we need something in place before launching most of our production samples.

I think our options are:
(1) Contribute directly on the LArSoft issue and push to have that resolved next week.
(2) Set `keepEMShowerDaughters` to True for SBND only, and perform a roll-up/truth-matching in the CAF creation.
(3) Some other clever solution?

**Comment by wesketchum** [comment]
Pinging @henrylay97 @jzennamo @marcodeltutto @cjbackhouse @miquelnebot @pgreen135 etc. here so you are aware.

I hate to do it, but I think actually it makes sense to change `keepEMShowerDaughters` to True in SBND for as part of production testing for next week (still need downstream for CAFs). If that is sensible, @marcodeltutto or @henrylay97, can I ask you to get a PR in with that change made prior to a patch 2021C release going out this week?

**Comment by jzennamo** [comment]
Could we create a second set of SBND fcl with `keepEMShowerDaughters == true` that we can use in test samples before we launch production?

**Comment by henrylay97** [comment]
So there already exists a fcl `g4_no_shower_rollup.fcl` which lives [here](https://github.com/SBNSoftware/sbndcode/blob/develop/sbndcode/JobConfigurations/standard/g4/g4_no_shower_rollup.fcl) which runs standard g4 without rollup.

On [this](https://github.com/SBNSoftware/sbndcode/tree/feature/hlay_no_rollup) branch I've added an equivalent for sce enabled g4. 

I'm currently running some tests on the standard CI to do a basic assessment of memory and time differences.

**Comment by henrylay97** [comment]
First set of CI results. Running the nu+cosmics regression test. The g4 stage (processing an artroot file with 2 BNB+Cosmic events).

[**v09_37_01**](https://dbweb8.fnal.gov:8443/LarCI/app/ns:sbnd/build_detail/test_details?build_id=sbnd_ci/8081&platform=Linux%203.10.0-1160.45.1.el7.x86_64&buildtype=slf7%20e20:prof&phase=ci_tests&test=nucosmics_g4_quick_test_sbndcode)
Elapsed time: 660.29 (s)
MaxRSS: 304.46 (10s of kb)
Scaled CPU: 3598.15 (kVax/s)

[**v09_37_01-NoRollUp**](https://dbweb8.fnal.gov:8443/LarCI/app/ns:sbnd/build_detail/test_details?build_id=sbnd_ci/8082&platform=Linux%203.10.0-1160.45.1.el7.x86_64&buildtype=slf7%20e20:prof&phase=ci_tests&test=nucosmics_g4_quick_test_sbndcode)
Elapsed time: 1928.84 (s)
MaxRSS: 524.71 (10s of kb)
Scaled CPU: 10563.17 (kVax/s)

Sorry for the weird units, just what the CI spits out. It gives clear proportional comparisons. Working on the effect of this change on subsequent stages (detsim->reco2) but this takes longer as it needs the sequential tests.

**Comment by henrylay97** [comment]
Further update, the largest time difference is coming from the `mcreco` stage not the actual `largeant` stage. There are also effects on file writing and pdfastsim. Do we need `mcreco`?

Before
```
1423: ================================================================================================================================
1424: TimeTracker printout (sec)                        Min           Avg           Max         Median          RMS         nEvts   
1425: ================================================================================================================================
1426: Full event                                      261.822       315.551       369.279       315.551       53.7284         2     
1427: --------------------------------------------------------------------------------------------------------------------------------
1428: source:RootInput(read)                        0.00118532    0.00303008    0.00487483    0.00303008    0.00184476        2     
1429: simulate:rns:RandomNumberSaver                8.1291e-05    0.000355157   0.000629024   0.000355157   0.000273867       2     
1430: simulate:loader:PhysListLoader                 3.952e-06    1.7116e-05     3.028e-05    1.7116e-05    1.3164e-05        2     
1431: simulate:largeant:larg4Main                     17.5651       19.1303       20.6955       19.1303       1.56521         2     
1432: simulate:ionandscint:IonAndScint               0.0745986      0.08271      0.0908215      0.08271     0.00811142        2     
1433: simulate:pdfastsim:PDFastSimPAR                 223.401       280.46        337.519       280.46        57.0587         2     
1434: simulate:simdrift:SimDriftElectrons             12.1172       13.8436       15.5699       13.8436       1.72638         2     
1435: simulate:mcreco:MCReco                         0.472245      0.479478      0.486711      0.479478     0.00723345        2     
1436: simulate:genericcrt:GenericCRT                 0.0616145     0.102518      0.143421      0.102518      0.0409034        2     
1437: [art]:TriggerResults:TriggerResultInserter     2.319e-05    4.9033e-05    7.4876e-05    4.9033e-05    2.5843e-05        2     
1438: end_path:out1:RootOutput                       8.462e-06    2.08435e-05   3.3225e-05    2.08435e-05   1.23815e-05       2     
1439: end_path:out1:RootOutput(write)                 1.43135       1.44562       1.45988       1.44562      0.0142622        2     
1440: ================================================================================================================================
```

After
```
1423: ================================================================================================================================
1424: TimeTracker printout (sec)                        Min           Avg           Max         Median          RMS         nEvts   
1425: ================================================================================================================================
1426: Full event                                      923.458       941.974       960.49        941.974       18.5155         2     
1427: --------------------------------------------------------------------------------------------------------------------------------
1428: source:RootInput(read)                        0.00223015    0.00676208     0.011294     0.00676208    0.00453193        2     
1429: simulate:rns:RandomNumberSaver                0.000544009   0.00220435    0.00386468    0.00220435    0.00166034        2     
1430: simulate:loader:PhysListLoader                 4.784e-06    2.41865e-05   4.3589e-05    2.41865e-05   1.94025e-05       2     
1431: simulate:largeant:larg4Main                     20.5834       22.6506       24.7177       22.6506       2.06711         2     
1432: simulate:ionandscint:IonAndScint               0.0803436     0.097878      0.115412      0.097878      0.0175343        2     
1433: simulate:pdfastsim:PDFastSimPAR                 262.645       333.594       404.543       333.594       70.949          2     
1434: simulate:simdrift:SimDriftElectrons             12.5029       14.5013       16.4998       14.5013       1.99846         2     
1435: simulate:mcreco:MCReco                          479.125       562.676       646.226       562.676       83.5506         2     
1436: simulate:genericcrt:GenericCRT                 0.0678691     0.115368      0.162868      0.115368      0.0474992        2     
1437: [art]:TriggerResults:TriggerResultInserter    4.5502e-05    6.7014e-05    8.8526e-05    6.7014e-05    2.1512e-05        2     
1438: end_path:out1:RootOutput                       9.621e-06    1.9139e-05    2.8657e-05    1.9139e-05     9.518e-06        2     
1439: end_path:out1:RootOutput(write)                 6.52578       8.31852       10.1113       8.31852       1.79274         2     
1440: ================================================================================================================================

**Comment by henrylay97** [comment]
So I've now tested four options
- Standard
- No rollup
- No mcreco (for reference)
- Neither rollup or mcreco (potential production version)

These are the results showing the effect on g4 AND on the following stages detsim reco1 and reco2.
_Note the difference between g4 quick and g4 seq is that the quick tests run 2 events whilst seq run 5._

![image](https://user-images.githubusercontent.com/68423730/143559900-6f010b8f-f1f6-48ea-aba9-d46c730996b7.png)

I know this is just tonnes of numbers now which isn't overly helpful. 

**TL;DR**
Turning off roll up has a ~3x increase on processing time and CPU for the g4 stage. Most of this can be removed with turning `mcreco` off (no following stages depend on it). There is still a marginal increase but its manageable I think.

The effect on detsim is negligible.

The effect on reco1 is more significant, close to a doubling in the time. The largest contributor to this is `gaushitTruthMatch` which seems logical. There also seems to be a large effect not accounted for in the per-event breakdown (accessing the file maybe?)

The effect on reco2 is in between (~1/3 increase in time). None of this appears in the per-event breakdown (again accessing the file?).

**Comment by pgreen135** [comment]
@miquelnebot this remains open, and critical -- the merged PR only contains some new fhicls for testing, we still do not have a decided upon solution

**Comment by miquelnebot** [comment]
> @miquelnebot this remains open, and critical -- the merged PR only contains some new fhicls for testing, we still do not have a decided upon solution / have not yet implemented this proposed solution as the default

👍 I meant regarding the SBN2021C production board, as things that are still needed. Is it?

**Comment by pgreen135** [comment]
yes, this is essential/blocking for SBN2021C for SBND

**Comment by jzennamo** [comment]
Does anyone have any pushback to this plan?

**Comment by pgreen135** [comment]
That sounds reasonable, it would be good to double check the output from MCReco without rollup is still as needed for these analyses (the 100x slowdown might point to some issue?)

**Comment by henrylay97** [comment]
> That sounds reasonable, it would be good to double check the output from MCReco without rollup is still as needed for these analyses (the 100x slowdown might point to some issue?)

yeah @pgreen135 makes a good point. [This](https://github.com/SBNSoftware/sbndcode/issues/230#issuecomment-979268223) comment contains the product differences and there are ~10x the number of MCShowers. This points to a fragmentation of showers that may be damaging to these analyses.

**Comment by jzennamo** [comment]
Agreed, once I have a test CAF file I was going to share it with the analysis teams

**Comment by jzennamo** [comment]
> That sounds reasonable, it would be good to double check the output from MCReco without rollup is still as needed for these analyses (the 100x slowdown might point to some issue?)

@pgreen135 where is there a 100x slowdown? I think I missed that message... :gulp:

**Comment by henrylay97** [comment]
I think Patrick is referring to the `mcreco` element of the g4 differences. MCReco went from ~.5s per event to ~500s per event (~1000x).

**Comment by jzennamo** [comment]
gotcha, MCReco has always been a "problem child" making it slow when there are many MCParticles, which is why it is nice to have the roll-up working. I think that this scale-up is expected since there are so many "subparticles" in an EM shower (sadly)

**Comment by pgreen135** [comment]
Closing now https://github.com/SBNSoftware/sbndcode/pull/234 has been merged.

**Comment by henrylay97** [comment]
Can I request we keep this open until someone has produced a CAF level solution to the other half of this issue?

**Comment by pgreen135** [comment]
sure, sorry @henrylay97 --- yes that is still outstanding, who is currently working on this? The CAF side of things is non-blocking for production to start right?

**Comment by henrylay97** [comment]
Depends whether we want to have to re-run the CAF stage I suppose! @jzennamo @wesketchum @cjbackhouse might be able to comment more on this side of things...

**Comment by pgreen135** [comment]
@henrylay97 is this now resolved / can be closed?

**Comment by henrylay97** [comment]
Yep! Not resolved but the original issue has gone / is going away so no longer relevant :)
